File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/networking/CanvasUpdatePacket.java
package pw.switchcraft.plethora.gameplay.modules.glasses.networking;

import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.PacketSender;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.network.PacketByteBuf;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject;
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasClient;
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectRegistry;

import java.util.ArrayList;
import java.util.List;

public class CanvasUpdatePacket {
    private int canvasId;
    private List<BaseObject> changed;
    private int[] removed;

    public CanvasUpdatePacket(int canvasId, List<BaseObject> changed, int[] removed) {
        this.canvasId = canvasId;
        this.changed = changed;
        this.removed = removed;
    }

    public CanvasUpdatePacket() {}

    public void fromBytes(PacketByteBuf buf) {
        canvasId = buf.readInt();

        int changedLength = buf.readInt();
        List<BaseObject> changed = this.changed = new ArrayList<>(changedLength);
        for (int i = 0; i < changedLength; i++) {
            changed.add(ObjectRegistry.read(buf));
        }

        // We sort on ID in order to guarantee parents are loaded before their children
        changed.sort(BaseObject.SORTING_ORDER);

        int removedLength = buf.readInt();
        int[] removed = this.removed = new int[removedLength];
        for (int i = 0; i < removedLength; i++) {
            removed[i] = buf.readInt();
        }
    }

    public void toBytes(PacketByteBuf buf) {
        buf.writeInt(canvasId);

        buf.writeInt(changed.size());
        for (BaseObject object : changed) {
            ObjectRegistry.write(buf, object);
        }

        buf.writeInt(removed.length);
        for (int id : removed) {
            buf.writeInt(id);
        }
    }

    public PacketByteBuf toBytes() {
        PacketByteBuf buf = PacketByteBufs.create();
        toBytes(buf);
        return buf;
    }

    public static void onReceive(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf,
                                 PacketSender responseSender) {
        CanvasUpdatePacket packet = new CanvasUpdatePacket();
        packet.fromBytes(buf);

        CanvasClient canvas = CanvasHandler.getClient(packet.canvasId);
        if (canvas == null) return;

        synchronized (canvas) {
            for (BaseObject obj : packet.changed) canvas.updateObject(obj);
            for (int id : packet.removed) canvas.remove(id);
        }
    }
}

File: src/main/java/io/sc3/plethora/gameplay/modules/glasses/networking/CanvasUpdatePacket.kt
package io.sc3.plethora.gameplay.modules.glasses.networking

import io.sc3.library.networking.ScLibraryPacket
import io.sc3.plethora.Plethora.ModId
import io.sc3.plethora.gameplay.modules.glasses.canvas.CanvasHandler.getClient
import io.sc3.plethora.gameplay.modules.glasses.objects.BaseObject
import io.sc3.plethora.gameplay.modules.glasses.objects.ObjectRegistry.read
import io.sc3.plethora.gameplay.modules.glasses.objects.ObjectRegistry.write
import net.fabricmc.fabric.api.networking.v1.PacketSender
import net.minecraft.client.MinecraftClient
import net.minecraft.client.network.ClientPlayNetworkHandler
import net.minecraft.network.PacketByteBuf

data class CanvasUpdatePacket(
  var canvasId: Int = 0,
  var changed: MutableList<BaseObject> = mutableListOf(),
  var removed: IntArray = IntArray(0)
): ScLibraryPacket() {
  override val id = CanvasUpdatePacket.id

  override fun toBytes(buf: PacketByteBuf) {
    buf.writeInt(canvasId)
    buf.writeCollection(changed, ::write)
    buf.writeIntArray(removed)
  }

  companion object {
    @JvmField
    val id = ModId("canvas_update")

    @JvmStatic
    fun fromBytes(buf: PacketByteBuf): CanvasUpdatePacket {
      val canvasId = buf.readInt()
      val changed = buf.readCollection({ mutableListOf<BaseObject>() }, ::read)
      val removed = buf.readIntArray()

      changed.sortWith(BaseObject.SORTING_ORDER) // Sort by ID to guarantee parents are loaded before their children

      return CanvasUpdatePacket(canvasId, changed, removed)
    }
  }

  override fun onClientReceive(client: MinecraftClient, handler: ClientPlayNetworkHandler,
                               responseSender: PacketSender) {
    val canvas = getClient(canvasId) ?: return
    synchronized(canvas) {
      changed.onEach(canvas::updateObject)
      removed.onEach(canvas::remove)
    }
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as CanvasUpdatePacket

    if (canvasId != other.canvasId) return false
    if (changed != other.changed) return false
    if (!removed.contentEquals(other.removed)) return false

    return true
  }

  override fun hashCode(): Int {
    var result = canvasId
    result = 31 * result + changed.hashCode()
    result = 31 * result + removed.contentHashCode()
    return result
  }
}
